integration-test/helpers/objectSchema.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * Returns the schema of a given JavaScript Object in the given Page. Schema is
+ * similar to the output of Object.getOwnPropertyDescriptors.
+ * Notes:
+ *  - Does not traverse up the target Object's __proto__ chain. If that is
+ *    desired, call `getObjectSchema` again for the Object's prototype.
+ *  - Functions are represented as a string "function(N)", where N is the number
+ *    of arguments that the function accepts (aka the function's length).
+ *  - Object keys are sorted alphabetically.
+ *  - Object values are handled recursively.
+ *  - Cyclic Object values are represented as the string "CYCLIC OBJECT VALUE".
+ * @param {Page} page
+ *   The Puppeteer page to inspect.
+ * @param {string} targetObjectName
+ *   The name of the Object to inspect. A string which when eval'd in the page
+ *   returns the target Object. For example, "window.someApi".
+ * @returns {Object}
+ *   The target Object's schema.
+ */
+async function getObjectSchema (page, targetObjectName) {
+    // eslint-disable-next-line no-eval
+    const targetObjectHandle = await page.evaluateHandle(name => eval(name), targetObjectName)
+
+    return await page.evaluate(function getObjectSchema (targetObject, previouslySeenObjects) {
+        const result = Object.create(null)
+        const descriptors = Object.getOwnPropertyDescriptors(targetObject)
+
+        // Keep track of Object values that we've already seen in order to avoid
+        // following cyclic Object values.
+        previouslySeenObjects = new Set(previouslySeenObjects)
+        previouslySeenObjects.add(targetObject)
+
+        const seenObjects = new Set(previouslySeenObjects)
```

Comment:
hm, I don't quite follow this code. why are we overriding previouslySeenObjects just to copy them right after?

BTW this code is getting sufficiently complex to have its own test set :) I don't want to derail you PR completely but have you considered using a library instead? https://www.npmjs.com/package/stringify-object

Reply:
yeah, fair enough, I also avoid adding 3p libraries where possible.

One more question about the code - I understand the purpose you mentioned, but what I don't understand is making a copy of the set twice:

> previouslySeenObjects = new Set(previouslySeenObjects)
> const seenObjects = new Set(previouslySeenObjects)

Reply:
Ah, good point! We don't need to do that, fixed. While at it, I added some more comprehensive comments.